### ISSUE TITLE
Normalization with mixed precision

### ISSUE THREAD

**Issue by javirk** [pytorch/pytorch#61954]
## 🐛 Bug

I am trying to use mixed precision with `autocast` and I am facing stability issues with the `nn.functional.normalize` function, returning NaN values when the corresponding float32 tensor works fine. All the steps can be reproduced without `autocast`, just making all tensors of type `torch.half`. In Colab the result of normalizing a very small float16 tensor is a kernel restart.

I have examined `nn.functional.normalize` function and I think the issue comes from the `clamp_min` part, which doesn't work for `torch.float16` tensors.

## To Reproduce

```
import torch
import torch.nn as nn

small = torch.tensor([1e-4, 1e-4, 2e-4], dtype=torch.float16)

# The following doesn't work
small_norm = nn.functional.normalize(small, dim=0)

# The following does work
small_norm = nn.functional.normalize(small.float(), dim=0)
```

To reproduce the behaviour with clamp function:
```
import torch

small = torch.tensor([1e-4, 1e-4, 2e-4], dtype=torch.float16)

# The following doesn't work
torch.clamp(small, min=1e-12)

# The following does work
torch.clamp(small.float(), min=1e-12)
```

## Expected behavior

`nn.functional.normalize` is in the [list](https://pytorch.org/docs/stable/amp.html#ops-that-can-autocast-to-float32) of operations that can autocast to float32, so I would expect no stability issues. I think it could be solved by changing the line

`denom = input.norm(p, dim, keepdim=True).clamp_min(eps).expand_as(input)`
to 
`denom = input.norm(p, dim, keepdim=True).float().clamp_min(eps).expand_as(input)`
(and the same in the else part). But I have no idea if this interferes with anything.

## Environment
 - PyTorch Version (e.g., 1.0): 1.9.0+cu102 (Google Colab default)
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.7.11
 - CUDA/cuDNN version: 11.2 (Google Colab)
 - GPU models and configuration: Tesla T4 (Colab) and RTX3090

cc @mcarilli @ptrblck

**Comment by vadimkantorov** [comment]
maybe normalize internally uses too small of an epsilon...

**Comment by javirk** [comment]
Sorry, I realized that the above code doesn't reproduce the issue, now I have it only during training with `autocast`. I close the issue and I will reopen it if I find a way to reproduce it.
